paperqa/agents/main.py

```diff
@@ -325,39 +339,31 @@ async def run_ldp_agent(
     **env_kwargs,
 ) -> tuple[PQASession, AgentStatus]:
     env = env_class(query, docs, **env_kwargs)
+    status = AgentStatus.SUCCESS
     # NOTE: don't worry about ldp import checks, because we know Settings.make_ldp_agent
     # has already taken place, which checks that ldp is installed
 
-    try:
-        async with asyncio.timeout(query.settings.agent.timeout):
-            rollout_manager = RolloutManager(
-                agent,
-                callbacks=[
-                    ldp_callback_type(
-                        env,
-                        on_env_reset_callback,
-                        on_agent_action_callback,
-                        on_env_step_callback,
-                    )
-                ],
-            )
-            await rollout_manager.sample_trajectories(
-                environments=[env], max_steps=query.settings.agent.max_timesteps
-            )
-            status = AgentStatus.SUCCESS
-    except TimeoutError:
-        logger.warning(
-            f"Agent timeout after {query.settings.agent.timeout}-sec, just answering."
+    async def rollout() -> None:
+        rollout_manager = RolloutManager(
+            agent,
+            callbacks=[
+                ldp_callback_type(
+                    env,
+                    on_env_reset_callback,
+                    on_agent_action_callback,
+                    on_env_step_callback,
+                )
+            ],
         )
-        status = AgentStatus.TRUNCATED
-        generate_answer_tool = next(
-            filter(lambda x: x.info.name == GenerateAnswer.TOOL_FN_NAME, env.tools)
+        trajs = await rollout_manager.sample_trajectories(
+            environments=[env], max_steps=query.settings.agent.max_timesteps
         )
-        await generate_answer_tool._tool_fn(question=query.query, state=env.state)
-    except Exception:
-        logger.exception(f"Agent {agent} failed.")
-        status = AgentStatus.FAIL
-    return env.state.session, status
+        traj = trajs[0]
+        if traj.steps[-1].truncated:
+            nonlocal status
+            status = AgentStatus.TRUNCATED
```

Comment:
This is confusing - can you just return status in `rollout`? This modification is like three levels deep and I'm actually unclear on if it always works

Reply:
Believe it or not, locally I have this exact change in the works

Reply:
Okay, done. Good to see we are both still thinking